File: src/api/serializer/cafe.py
from rest_framework import serializers
from src.core.models import CafeRoom, MenuCategory, MenuItem, Table, Order, OrderItem, User, InventoryItem


class CafeRoomSerializer(serializers.ModelSerializer):
    class Meta:
        model = CafeRoom
        fields = ['id', 'location', 'name', 'width', 'height', 'position_x', 'position_y', 'is_active']


class MenuItemSerializer(serializers.ModelSerializer):
    image_url = serializers.ReadOnlyField()

    class Meta:
        model = MenuItem
        fields = ['id', 'category', 'name', 'description', 'price', 'image', 'image_url', 'is_available', 'order']


class MenuCategorySerializer(serializers.ModelSerializer):
    items = MenuItemSerializer(many=True, read_only=True)

    class Meta:
        model = MenuCategory
        fields = ['id', 'location', 'name', 'order', 'is_active', 'items']


class TableSerializer(serializers.ModelSerializer):
    status = serializers.SerializerMethodField()
    room_name = serializers.CharField(source='room.name', read_only=True)

    class Meta:
        model = Table
        fields = ['id', 'location', 'room', 'room_name', 'number', 'name', 'capacity', 'is_active', 'qr_token', 'position_x', 'position_y', 'shape', 'width', 'height', 'status']

    def get_status(self, obj):
        open_order = obj.orders.filter(status='open').first()
        if not open_order:
            return {'has_open_order': False, 'order_id': None, 'item_count': 0}
        return {
            'has_open_order': True,
            'order_id': open_order.id,
            'item_count': open_order.items.count(),
        }


class OrderItemSerializer(serializers.ModelSerializer):
    menu_item_name = serializers.ReadOnlyField()
    menu_item_price = serializers.ReadOnlyField()

    class Meta:
        model = OrderItem
        fields = ['id', 'order', 'menu_item', 'menu_item_name', 'menu_item_price', 'quantity', 'price', 'note']


class OrderSerializer(serializers.ModelSerializer):
    items = OrderItemSerializer(many=True, read_only=True)
    total = serializers.ReadOnlyField()
    table_name = serializers.ReadOnlyField()
    waiter_name = serializers.ReadOnlyField()
    status_display = serializers.CharField(source='get_status_display', read_only=True)
    order_type_display = serializers.CharField(source='get_order_type_display', read_only=True)

    class Meta:
        model = Order
        fields = [
            'id', 'table', 'table_name', 'waiter', 'waiter_name',
            'status', 'status_display', 'order_type', 'order_type_display',
            'customer_name', 'customer_phone', 'delivery_address',
            'total', 'note', 'opened_at', 'closed_at', 'items',
        ]


class InventoryItemSerializer(serializers.ModelSerializer):
    is_low = serializers.ReadOnlyField()

    class Meta:
        model = InventoryItem
        fields = ['id', 'location', 'name', 'unit', 'quantity', 'min_quantity', 'is_low']


class StaffSerializer(serializers.ModelSerializer):
    class Meta:
        model = User
        fields = ['id', 'username', 'full_name', 'role', 'location', 'is_active', 'date_joined']
        read_only_fields = ['id', 'date_joined']


class StaffCreateSerializer(serializers.ModelSerializer):
    password = serializers.CharField(write_only=True, min_length=4)

    class Meta:
        model = User
        fields = ['id', 'username', 'full_name', 'role', 'location', 'is_active', 'password']

    def validate_role(self, value):
        if value not in ('waiter', 'cafe_manager', 'kitchen'):
            raise serializers.ValidationError("Faqat waiter, cafe_manager yoki kitchen rol bo'lishi mumkin.")
        return value

    def create(self, validated_data):
        password = validated_data.pop('password')
        user = User(**validated_data)
        user.set_password(password)
        user.save()
        return user

File: frontend/src/api/cafe.ts
import apiClient from './client'

// ── Menu ──────────────────────────────────────────────────────────────────────
export const menuApi = {
  categories: (locationId?: number) =>
    apiClient.get('cafe/menu-categories/', { params: locationId ? { location_id: locationId } : {} }),
  createCategory: (data: { location: number; name: string; order?: number }) =>
    apiClient.post('cafe/menu-categories/', data),
  updateCategory: (id: number, data: Partial<{ name: string; order: number; is_active: boolean }>) =>
    apiClient.patch(`cafe/menu-categories/${id}/`, data),
  deleteCategory: (id: number) => apiClient.delete(`cafe/menu-categories/${id}/`),

  items: (params?: { location_id?: number; category_id?: number }) =>
    apiClient.get('cafe/menu-items/', { params }),
  createItem: (data: FormData | Record<string, unknown>) =>
    apiClient.post('cafe/menu-items/', data),
  updateItem: (id: number, data: FormData | Record<string, unknown>) =>
    apiClient.patch(`cafe/menu-items/${id}/`, data),
  deleteItem: (id: number) => apiClient.delete(`cafe/menu-items/${id}/`),
}

// ── Tables ────────────────────────────────────────────────────────────────────
export const tableApi = {
  list: (locationId?: number) =>
    apiClient.get('cafe/tables/', { params: locationId ? { location_id: locationId } : {} }),
  create: (data: { location: number; room?: number | null; number?: number; name?: string; capacity?: number; position_x?: number; position_y?: number; shape?: string; width?: number; height?: number }) =>
    apiClient.post('cafe/tables/', data),
  update: (id: number, data: Partial<{ room: number | null; name: string; capacity: number; position_x: number; position_y: number; shape: string; width: number; height: number }>) => apiClient.patch(`cafe/tables/${id}/`, data),
  delete: (id: number) => apiClient.delete(`cafe/tables/${id}/`),
}

export const roomApi = {
  list: (locationId?: number) => apiClient.get('cafe/rooms/', { params: locationId ? { location_id: locationId } : {} }),
  create: (data: { location: number; name: string; width?: number; height?: number; position_x?: number; position_y?: number }) => apiClient.post('cafe/rooms/', data),
  update: (id: number, data: Partial<{ name: string; width: number; height: number; position_x: number; position_y: number }>) => apiClient.patch(`cafe/rooms/${id}/`, data),
  delete: (id: number) => apiClient.delete(`cafe/rooms/${id}/`),
}

// ── Orders ────────────────────────────────────────────────────────────────────
export const orderApi = {
  list: (params?: { location_id?: number; status?: string; order_type?: string }) =>
    apiClient.get('cafe/orders/', { params }),
  today: (locationId?: number) =>
    apiClient.get('cafe/orders/today/', { params: locationId ? { location_id: locationId } : {} }),
  create: (data: {
    table?: number
    order_type?: string
    customer_name?: string
    customer_phone?: string
    delivery_address?: string
    note?: string
  }) => apiClient.post('cafe/orders/', data),
  addItems: (orderId: number, items: Array<{ menu_item: number; quantity: number; note?: string }>) =>
    apiClient.post(`cafe/orders/${orderId}/add_items/`, { items }),
  updateStatus: (orderId: number, status: string) =>
    apiClient.post(`cafe/orders/${orderId}/update_status/`, { status }),
}

// ── Locations ─────────────────────────────────────────────────────────────────
export const locationApi = {
  list: () => apiClient.get('locations/'),
}

// ── Inventory ─────────────────────────────────────────────────────────────────
export const inventoryApi = {
  list: (locationId?: number) =>
    apiClient.get('cafe/inventory/', { params: locationId ? { location_id: locationId } : {} }),
  create: (data: { location: number; name: string; unit: string; quantity: number; min_quantity: number }) =>
    apiClient.post('cafe/inventory/', data),
  update: (id: number, data: Partial<{ name: string; unit: string; quantity: number; min_quantity: number }>) =>
    apiClient.patch(`cafe/inventory/${id}/`, data),
  delete: (id: number) => apiClient.delete(`cafe/inventory/${id}/`),
  adjust: (id: number, delta: number) => apiClient.post(`cafe/inventory/${id}/adjust/`, { delta }),
}

// ── Analytics ─────────────────────────────────────────────────────────────────
export const analyticsApi = {
  cafe: (params?: { days?: number; location_id?: number }) =>
    apiClient.get('cafe/analytics/', { params }),
}

// ── Staff ─────────────────────────────────────────────────────────────────────
export const staffApi = {
  list: () => apiClient.get('cafe/staff/'),
  create: (data: { username: string; full_name: string; role: string; password: string; is_active?: boolean }) =>
    apiClient.post('cafe/staff/', data),
  update: (id: number, data: Partial<{ full_name: string; role: string; is_active: boolean }>) =>
    apiClient.patch(`cafe/staff/${id}/`, data),
  delete: (id: number) => apiClient.delete(`cafe/staff/${id}/`),
  setPassword: (id: number, password: string) =>
    apiClient.post(`cafe/staff/${id}/set_password/`, { password }),
  toggleActive: (id: number) => apiClient.post(`cafe/staff/${id}/toggle_active/`),
}
